ROX-17670: prune plop with SQL instead of a walk

dashrews78 opened the pull request:
## Description

This change does the following:
- promotes closed timestamp field to a column as it was used in pruning via a walk
- updated pruning to use SQL instead of a walk
- Added migration to populate the closed timestamp column if the field is not null
- If the field is null the column will be null and that data will not be pruned
- Updated the unit tests for pruning

This change was just as much about using good access patterns as it was about performance.  Looking at each record to determine if we should delete it or not when the database can do it in one command is a poor pattern.  In light of that, here are some numbers.  I ran simple go code to add 2M PLOPS and to prune them.  

Old way with walk and delete by ids took around 22-23 seconds:
```
start removing lots of PLOP 2023-09-22 09:41:26.48457 -0400 EDT m=+438.911355148
done removing lots of PLOP 2023-09-22 09:41:48.672003 -0400 EDT m=+461.098773543

start removing lots of PLOP 2023-09-22 09:48:32.790618 -0400 EDT m=+865.217130555
done removing lots of PLOP 2023-09-22 09:48:55.365282 -0400 EDT m=+887.791779782

start removing lots of PLOP 2023-09-22 09:55:17.979734 -0400 EDT m=+1270.405987111
done removing lots of PLOP 2023-09-22 09:55:40.191454 -0400 EDT m=+1292.617693236
```

New way with single delete SQL statement took around 15 seconds:
```
start prune 2023-09-22 08:46:51.11986 -0400 EDT m=+64.231366233
end prune 2023-09-22 08:47:05.53358 -0400 EDT m=+78.645076711

start prune 2023-09-22 08:53:53.685097 -0400 EDT m=+486.796332463
end prune 2023-09-22 08:54:08.868958 -0400 EDT m=+501.980184616

start prune 2023-09-22 09:00:24.826337 -0400 EDT m=+877.937322993
end prune 2023-09-22 09:00:39.526209 -0400 EDT m=+892.637185432
```

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
~- [ ] Evaluated and added CHANGELOG entry if required~
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

Went through upgrade and rollback tests manually.
- Started 4.1.3
- upgraded to image in this PR ensuring migration ran and new column was created
- downgraded back to 4.1.3 ensuring column remained and software continued to function
- upgraded back to PR in this image ensuring migration executed again.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

dashrews78 committed:
X-Smart-Branch-Parent: master

dashrews78 commented:
/retest

JoukoVirtanen commented:
Since this is supposed to improve performance can you show some metrics demonstrating that this improves performance?

dashrews78 commented:
> Since this is supposed to improve performance can you show some metrics demonstrating that this improves performance?

I thought about that but that may be tough to do.  We don't have a good current metric to use.  And the other thing is part of this is removing the anti-pattern of pulling all the data into code, looking at it, and then deleting a list of ids.  

I will give it some thought.

JoukoVirtanen commented:
One problem with pruning PLOP is that orphaned listening endpoints where the pod has been deleted are not pruned. That could be done in another PR, but if you would like to do it in this PR, that would be great.

JoukoVirtanen commented:
> > Since this is supposed to improve performance can you show some metrics demonstrating that this improves performance?
> 
> I thought about that but that may be tough to do. We don't have a good current metric to use. And the other thing is part of this is removing the anti-pattern of pulling all the data into code, looking at it, and then deleting a list of ids.
> 
> I will give it some thought.

I did performance testing for PLOP when it was first being worked on. I created documents for the performance testing. Here is an out of date document on some of the testing that I did https://docs.google.com/document/d/1Ud8DFhS6qDsMDI2N49jKqj6gEX79yGyXF7QOcVIBVuo/edit

The performance testing scripts I used have been moved to https://github.com/stackrox/performance-scale-tests. Others have done performance testing by using kube-burner more directly.

connorgorman commented on central/postgres/pruning.go:
```diff
@@ -87,6 +87,9 @@ const (
 	pruneLogImbues = `DELETE FROM log_imbues WHERE timestamp < now() at time zone 'utc' - INTERVAL '%d MINUTES'`
+
+	// Delete orphaned PLOPs
+	pruneOrphanedPLOPs = `DELETE FROM listening_endpoints WHERE closetimestamp < now() at time zone 'utc' - INTERVAL '%d MINUTES'`
```
Not sure that I follow that this is a orphaned plop? @JoukoVirtanen Do we not keep history about which plops were opened if the pod is still around?

JoukoVirtanen replied:
The only criteria for pruning is if the listening endpoint has been closed longer than some specified amount of time. We also need to add pruning for when the pod has been deleted, but the listening endpoint has not been reported as being closed. What criteria are you suggesting?

JoukoVirtanen replied:
I think improving the pruning criteria can be done in another PR.

dashrews78 replied:
@JoukoVirtanen Fixing the PLOP criteria should probably be tracked with a ticket in addition to a PR.

JoukoVirtanen replied:
I have created a ticket here https://issues.redhat.com/browse/ROX-19833

There was already a ticket for not reporting listening endpoints that whose pods have been deleted https://issues.redhat.com/browse/ROX-19577

I have a draft PR here https://github.com/stackrox/stackrox/pull/7858

I don't think I am doing things the best way there yet. I need to write some thoughts down and do some testing.

connorgorman commented on migrator/migrations/m_189_to_m_190_plop_add_closed_time_column/migration_impl.go:
```diff
@@ -0,0 +1,70 @@
+	log       = logging.LoggerForModule()
+)
+
+func migrate(database *types.Databases) error {
```
Do you have a ballpark of how long this will take? I would test with at least something like half a million and see how long this takes

dashrews78 replied:
Ran 2M in a unit test.  Took 2:47

connorgorman replied:
Ok, that's not so bad 👍 thanks for the test

connorgorman commented on migrator/migrations/m_189_to_m_190_plop_add_closed_time_column/migration_impl.go:
```diff
@@ -0,0 +1,70 @@
+	if err != nil {
+		return errors.Wrapf(err, "failed to iterate table %s", updatedSchema.ListeningEndpointsTableName)
+	}
+	if rows.Err() != nil {
```
idiomatic to do err :=, then we don't need to do rows.Err() repeatedly

dashrews78 replied:
Not 100% certain what you were getting at here.  But I moved this to the end of the loop.  Since gorm uses pgx underneath, the error may not surface in `rows.Err()` until some later iteration, so `rows.Err()` has to be checked after the loop.

JoukoVirtanen commented:
> > > Since this is supposed to improve performance can you show some metrics demonstrating that this improves performance?
> > 
> > 
> > I thought about that but that may be tough to do. We don't have a good current metric to use. And the other thing is part of this is removing the anti-pattern of pulling all the data into code, looking at it, and then deleting a list of ids.
> > I will give it some thought.
> 
> I did performance testing for PLOP when it was first being worked on. I created documents for the performance testing. Here is an out of date document on some of the testing that I did https://docs.google.com/document/d/1Ud8DFhS6qDsMDI2N49jKqj6gEX79yGyXF7QOcVIBVuo/edit
> 
> The performance testing scripts I used have been moved to https://github.com/stackrox/performance-scale-tests. Others have done performance testing by using kube-burner more directly.

This is probably overkill, but it would still be nice to see some kind of metric on this. Maybe generate a bunch of fake data for a pruning unit test and measure the time for both methods.

dashrews78 committed:
X-Smart-Squash: Squashed 6 commits:

14844ec549 prune plop
552ad16425 style
6b21a84061 style
ba045f1bf3 some comment updates
8013d8bb29 lint probs
4cb4f25e75 move check to rows.Err

dashrews78 committed:
rebase due to other migration coming in before I finished

dashrews78 committed:
missed all

dashrews78 committed:
style

dashrews78 commented:
/retest

connorgorman commented on migrator/migrations/m_190_to_m_191_plop_add_closed_time_column/migration_test.go:
```diff
@@ -0,0 +1,84 @@
+	s.Require().Equal(numPlop, n)
+
+	// Now run the migration
+	log.Infof("SHREWS start -- %v", time.Now())
```
remove

connorgorman reviewed (changes requested):
Minor things, but need to remove extra logs plus use the helper functions from require

dashrews78 committed:
clean up and use Contains

porridge commented:
I think this might have caused https://issues.redhat.com/browse/ROX-20017
